src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -3,108 +3,49 @@ package storage_power_consensus
-	case sector.DetectedFault:
-		return 10 // placeholder
-	case sector.TerminatedFault:
-		return 100 // placeholder
-	default:
-		panic("Case not supported")
-	}
-}
-
-////////////////////////////////////////////////////////////////////////////////
-// Boilerplate
 ////////////////////////////////////////////////////////////////////////////////
-type InvocOutput = vmr.InvocOutput
-type Runtime = vmr.Runtime
-type Bytes = util.Bytes
-type State = StoragePowerActorState
-
-func (a *StoragePowerActorCode_I) State(rt Runtime) (vmr.ActorStateHandle, State) {
-	h := rt.AcquireState()
-	stateCID := h.Take()
-	stateBytes := rt.IpldGet(ipld.CID(stateCID))
-	if stateBytes.Which() != vmr.Runtime_IpldGet_FunRet_Case_Bytes {
-		rt.AbortAPI("IPLD lookup error")
-	}
-	state := DeserializeState(stateBytes.As_Bytes())
-	return h, state
-}
-func Release(rt Runtime, h vmr.ActorStateHandle, st State) {
-	checkCID := actor.ActorSubstateCID(rt.IpldPut(st.Impl()))
-	h.Release(checkCID)
-}
-func UpdateRelease(rt Runtime, h vmr.ActorStateHandle, st State) {
-	newCID := actor.ActorSubstateCID(rt.IpldPut(st.Impl()))
-	h.UpdateRelease(newCID)
-}
-func (st *StoragePowerActorState_I) CID() ipld.CID {
-	panic("TODO")
-}
-func DeserializeState(x Bytes) State {
-	panic("TODO")
-}
-
+// Actor methods
 ////////////////////////////////////////////////////////////////////////////////
 
-func (a *StoragePowerActorCode_I) AddBalance(rt Runtime) {
-	rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_Account)
+func (a *StoragePowerActorCode_I) AddBalance(rt Runtime, minerAddr addr.Address) {
+	if !rt.ValidateIsStorageMiner(minerAddr) {
+		rt.AbortArgMsg("spa.AddBalance: invalid minerAdder.")
+	}
+	// either worker or owner can add balance to a miner
+	rt.ValidateImmediateCallerDetermineRecipient(minerAddr)
 
-	ownerAddr := rt.ImmediateCaller()
 	msgValue := rt.ValueReceived()
 
 	h, st := a.State(rt)
-	newTable, ok := actor.BalanceTable_WithAdd(st.EscrowTable(), ownerAddr, msgValue)
+	newTable, ok := actor.BalanceTable_WithAdd(st.EscrowTable(), minerAddr, msgValue)
 	if !ok {
 		rt.AbortStateMsg("spa.AddBalance: Escrow operation failed.")
 	}
 	st.Impl().EscrowTable_ = newTable
 	UpdateRelease(rt, h, st)
 }
 
-func (a *StoragePowerActorCode_I) WithdrawBalance(rt Runtime, amountRequested actor.TokenAmount) {
-	rt.ValidateImmediateCallerAcceptAnyOfType(actor.BuiltinActorID_Account)
+func (a *StoragePowerActorCode_I) WithdrawBalance(rt Runtime, minerAddr addr.Address, amountRequested actor.TokenAmount) {
 
 	if amountRequested < 0 {
 		rt.AbortArgMsg("spa.WithdrawBalance: negative amount.")
 	}
 
-	minerAddr := rt.ImmediateCaller()
-
-	var ownerAddr addr.Address
-	TODO() // Determine owner address from miner
+	if !rt.ValidateIsStorageMiner(minerAddr) {
+		rt.AbortArgMsg("spa.WithdrawBalance: invalid minerAddr.")
+	}
+	// either worker or owner can withdraw balance from a miner
+	rt.ValidateImmediateCallerDetermineRecipient(minerAddr)
+	msgSender := rt.ImmediateCaller()
```

Comment:
The name `ValidateImmediateCallerDetermineRecipient` suggests that the return value should be used as the recipient. If we want different logic here then that points to it being a bad function to have on the Runtime since it's so special-case. Don't attempt to share the wrong function, write the correct one here.

Also I'm pretty sure we had decided in conversation that pledge collateral could only be withdrawn to the owner.